src/types/tdigest.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <fmt/format.h>
+
+#include <vector>
+
+#include "status.h"
+
+struct Centroid {
+  double mean;
+  double weight = 1.0;
+
+  // merge with another centroid
+  void Merge(const Centroid& centroid) {
+    weight += centroid.weight;
+    mean += (centroid.mean - mean) * centroid.weight / weight;
+  }
+
+  std::string ToString() const { return fmt::format("centroid<mean: {}, weight: {}>", mean, weight); }
+
+  template <typename T>
+  explicit Centroid(T&& centroid) : mean(centroid.mean), weight(centroid.weight) {}
+
+  explicit Centroid() = default;
+  explicit Centroid(double mean, double weight) : mean(mean), weight(weight) {}
+};
+
+struct CentroidsWithDelta {
+  std::vector<Centroid> centroids;
+  uint64_t delta;
+  double min;
+  double max;
+  double total_weight;
+};
+
+StatusOr<CentroidsWithDelta> TDigestMerge(const std::vector<CentroidsWithDelta>& centroids_list);
+StatusOr<CentroidsWithDelta> TDigestMerge(const std::vector<double>& buffer, const CentroidsWithDelta& centroid_list);
+
+/**
+
+TD should looks like below:
+class TDSample {
+  public:
+  struct Iterator {
+    Iterator* Clone() const;
+    bool Next();
+    bool Valid() const;
+    StatusOr<Centroid> GetCentroid() const;
+  };
+  Iterator* Begin();
+  Iterator* End();
+  double TotalWeight();
+  double Min() const;
+  double Max() const;
+};
+
+**/
+
+// a numerically stable lerp is unbelievably complex
+// but we are *approximating* the quantile, so let's keep it simple
+// reference:
+// https://github.com/apache/arrow/blob/27bbd593625122a4a25d9471c8aaf5df54a6dcf9/cpp/src/arrow/util/tdigest.cc#L38
+static inline double Lerp(double a, double b, double t) { return a + t * (b - a); }
+
+template <typename TD, typename Lerp>
+inline StatusOr<double> TDigestQuantile(TD&& td, double q, Lerp lerp) {
```

Comment:
why using template for `TD`? Also why `Lerp` is an argument?

Reply:
Hi @mapleFU ,

I want to make the `TD` an static implementation which can be evolved in future.
I try to reduce the virtual function, so I choose a template implementation.

The `Lerp` is an abstract function here, maybe it could be a dynamic runtime one like `std::function`.
So, I decide to use an argument here.

Please let me know if here can be improved. 😄

Best Regards,
Edward

Reply:
I mean would it able to give the different lerp function?

Reply:
Hi @mapleFU ,

Thanks for your suggestion! 😄

I have replaced `Lerp` template argument to `Lerp` function itself.
I think this function may not change frequently.

Best Regards,
Edward